### ISSUE TITLE
Reinforcement Learning

### ISSUE THREAD

**Issue by raddy** [fchollet/keras#167]
I've done zero work in RL so this might be REALLY stupid...

Is the cleanest method generally just to add an "Action" column to each matrix in your inputs and create a new objective function for your rewards? 

After mucking with this for a day, I found that I couldn't really get better performing models than just fitting an estimated reward straight to MAE from my state matrix. My predictions just approached that fantasy 'decision boundary' and started having ugly gradients.

Or am I massively misunderstanding how this is normally stacked. The other thought was to just "encourage" the model to predict values partitioned at 0 via the loss function. 

In the case where the action state space is discrete and just 0 or 1, where you either leave the lever alone or pull it I was working with a loss function along the lines:

```
def slot_reward(y_true,y_pred):
    y_pred = (T.sgn(y_pred)+1)/2
    return ((y_pred * y_true)*-1).sum()
```

This seems to blow up the gradients pretty badly too. 

Any advice?

**Comment by fchollet** [comment]
If you think your loss is sound (I don't have enough context to know whether that's actually the case) but your gradients are exploding throughout the network, a simple solution is to clip your gradients, which you can do by passing the argument `clipnorm=1` (or some other value) to your optimizer.

What I would do (disclaimer: I have thought about this for about 5s, so it might not be useful at all)... is consider the input of the network to be sensors, the output to be actuators (binarize it), and define the environment logic in the labels/targets. The targets would be an estimation of the reward/penalty for each action or action combination given a certain sensory input, and the loss function would simply associate the predicted action with its pre-computed reward/penalty and return that.

The above is purely stochastic (no memory) so it might not be what you want to do, though.

**Comment by fchollet** [comment]
I'm curious as to what you ended up doing...?

**Comment by raddy** [comment]
Still working on it. ^^

Generally been trying to use something like the previous scoring function vs a sigmoid 0-1 output representing not-pull or pull the lever. 

So far, I haven't had much luck getting my scoring function to guide much learning. (A few hand chosen rules on the feature set is easily outscoring the net) I'm getting pretty decent results out of simple gradient boosted trees, so I've been slacking a bit with this solution.

**Comment by c0d3rman** [comment]
Any progress? I'm trying to do RL with keras as well.

**Comment by laubosslink** [comment]
I'm currently intersting too.

**Comment by swayson** [comment]
+1

**Comment by Sebubu** [comment]
+1

**Comment by t3476** [comment]
For more general cases, I think pretending negative reward as loss should works.
    def negative_reward(y_true, y_pred):
        # pretend y_pred as actions, ignore y_true
        negative_reward = - get_reward_from_action(y_pred)

**Comment by fchollet** [comment]
@t3476 that would work only as long as `get_reward_from_action` is a Theano (or TensorFlow) expression. Which may be somewhat limiting.

**Comment by t3476** [comment]
I deleted and repost because of malformed code
For more general cases, I think pretending negative reward as loss should works.

    def negative_reward(y_true, y_pred):
       # pretend y_pred as actions, ignore y_true
       negative_reward = - get_reward_from_action(y_pred)

**Comment by t3476** [comment]
@fchollet 
I don't see why it's limiting since reward is just a scalar. If it is limiting, as a workaround, you can provide both version for Theano and Tensorflow and let function decides which one to be used by detecting loaded modules.

**Comment by neilslater** [comment]
I'm interested in this too - I don't think this is an issue with Keras at all, but would love to see a simple example of Keras used as the reward estimator model in an RL context.

**Comment by fchollet** [comment]
What specific tasks are you looking into applying this to? Is this something you would like to add to the Keras examples?

**Comment by neilslater** [comment]
A personal goal/project this year for me is to write RL that learns to play a simple multiplayer board or card game. For this and other projects (e.g. entering Kaggle competitions) I am currently researching a NN library to follow on from https://github.com/rasmusbergpalm/DeepLearnToolbox  and am probably going to settle on Keras from what I have read so far. 

Aside: I currently know very little Python, but it is clear that my professional language of Ruby is not going to gain any ML libraries that compete with it in the next year, and I am through with trying to write them (e.g. https://github.com/neilslater/ru_ne_ne will never be finished, let alone compete with *any* Python library).

Given enough time, if I succeed in my project, and there is still no example here, I will happily submit something as an added example in a PR for Keras.

**Comment by dandxy89** [comment]
Take a look at [this](https://github.com/dandxy89/rf_helicopter) - Helicopter Game using Q-Learning and Keras :+1:
